### PR TITLE
Update Pisa DNN, add DY ZpT reweighting, add signal strength parameter for DY in data card

### PR DESCRIPTION
**Pull request by nanlu06**
Changes include:

1. Update Pisa DNN, corresponding to Pisa group's report in the Hmm VBF meeting: https://indico.cern.ch/event/844259/?showDate=2019-09-10&showSession=45

**Next action item:
(1) check the N soft jet variable calculation and see if it is consistent with https://github.com/arizzi/PisaHmm/blob/1d532b59afde428ee38412c7673f1671459f28aa/eventprocessing.py#L89-L95
(2)and to address this:
Leonardo Giannini:
"The training is done using Herwig EWK Zjj samples that are also in the evaluation. In order not have a bias we have a different training to be applied to a subset of events based on the event number:
so evt0 files must be used for event%4 == 0, evt1 for event%4 == 1 and so on."**

2. implemented ZpT reweighing.

3. added two signal strengths parameters for DY in datacard so that to keep it uncorrelated for Z CR and others, commented out the 10% normalization uncertainty.  **The next action item is to add the uncertainty calculated using the LHE weights (shape).**

4. add one line of code in the datacard to enable MC stats uncertainty.

5. adding one variable nomweight in DNN vars which includes SF. This should be used to calculate event yield based on *.npy files

### CODE REVIEW

File: tests/hmm/hmumu_utils.py [hunk 1]
```diff
@@ -1188,6 +1204,38 @@ def get_bit_values(array, bit_index):
...
     return (array & 2**(bit_index)) >> 1
 
+# Custom kernels to get the number of softJets with pT>5 GEV
+def nsoftjets(nevt,softjets, leading_muon, subleading_muon, leading_jet, subleading_jet, ptcut, dr2cut):
+    njet_out = NUMPY_LIB.zeros(nevt, dtype=NUMPY_LIB.int32)
+    nsoftjets_cpu(nevt, softjets.offsets, softjets.pt, softjets.eta, softjets.phi, leading_jet["eta"], subleading_jet["eta"], leading_jet["phi"], subleading_jet["phi"], leading_muon["eta"], subleading_muon["eta"], leading_muon["phi"], subleading_muon["phi"], ptcut, dr2cut, njet_out)
+    return njet_out
+
+@numba.njit(parallel=True, fastmath=True)
+def nsoftjets_cpu(nevt, softjets_offsets, pt, eta, phi, etaj1, etaj2, phij1, phij2, etam1, etam2, phim1, phim2, ptcut, dr2cut, njet_out):
```
Comment (jpata) [inline tests/hmm/hmumu_utils.py]: If you want, you could also achieve deltaR matching using the special function `mask_deltar_first`, here is an example: https://github.com/jpata/hepaccelerate-cms/blob/master/tests/hmm/hmumu_utils.py#L130.

File: tests/hmm/analysis_hmumu.py [hunk 2]
```diff
@@ -335,17 +335,19 @@ def __init__(self, args, do_tensorflow=True, gpu_memory_fraction=0.2):
...
-                self.dnnPisa_model = keras.models.model_from_json(file_handle.read())
-            self.dnnPisa_model.load_weights("data/PisaDNN/model_preparation/model_toexport_evt0.h5")
-            self.dnnPisa_normfactors1 = np.load("data/PisaDNN/model_preparation/helphelp_node1.npy")
-            self.dnnPisa_normfactors2 = np.load("data/PisaDNN/model_preparation/helphelp_node2.npy")
-            if args.use_cuda:
-                import cupy
-                self.dnnPisa_normfactors1 = cupy.array(self.dnnPisa_normfactors1[0]), cupy.array(self.dnnPisa_normfactors1[1])
-                self.dnnPisa_normfactors2 = cupy.array(self.dnnPisa_normfactors2[0]), cupy.array(self.dnnPisa_normfactors2[1])
+            
+            for imodel in range(4):
```
Comment (jpata) [inline tests/hmm/analysis_hmumu.py]: to make maintaining the code easier in the future, I would suggest to take this `4` from the parameters dictionary, as it is needed also later in the code.

File: tests/hmm/hmumu_utils.py [hunk 3]
```diff
@@ -436,12 +447,14 @@ def analyze_data(
                 dnn_vars["lumi"] = scalars["luminosityBlock"][dnn_presel]
                 dnn_vars["event"] = scalars["event"][dnn_presel]
                 dnn_vars["dnn_pred"] = dnn_prediction
-                dnn_vars["dnnPisa_pred"] = dnnPisa_prediction
+                if not (len(dnnPisa_predictions)==0):
+                    for imodel in range(4):
```
Comment (jpata) [inline tests/hmm/hmumu_utils.py]: here you could also get this `4` from parameters.

File: tests/hmm/hmumu_utils.py [hunk 4]
```diff
@@ -436,12 +447,14 @@ def analyze_data(
...
+                        dnn_vars["dnnPisa_pred"+str(imodel)] = dnnPisa_predictions[imodel]
                 #Save the DNN training ntuples as npy files
                 if parameters["save_dnn_vars"] and jet_syst_name[0] == "nominal" and parameter_set_name == "baseline":
                     dnn_vars_np = {k: NUMPY_LIB.asnumpy(v) for k, v in dnn_vars.items()}
                     if is_mc:
-                        dnn_vars_np["nomweight"] = NUMPY_LIB.asnumpy(weights_in_dnn_presel["nominal"]*genweight_scalefactor)
+                        dnn_vars_np["nomweight"] = NUMPY_LIB.asnumpy(weights_in_dnn_presel["nominal"]/genweight_scalefactor)
```
Comment (jpata) [inline tests/hmm/hmumu_utils.py]: Typo changing `*` to `/` or intended?

**Comment by jpata** [comment]
Thank you for making a PR. The description is in line with the changes. I had a few smaller suggestions, the main question is changing `*` to `/` for the `genweight_scalefactor`.

**Comment by jpata** [comment]
@nanlu06 can you check this PR? Seems like you need to merge to master due to the datacard updates. After this, can you let me know if you would like me to test this, or feel free to merge yourself if you are satisfied with the tests.

File: tests/hmm/hmumu_utils.py (hunk 3 above)
Comment (nanlu06) [reply tests/hmm/hmumu_utils.py]: removed the hard-coded "4"

File: tests/hmm/hmumu_utils.py (hunk 1 above)
Comment (nanlu06) [reply tests/hmm/hmumu_utils.py]: Yes, will try to use this function later.

**Comment by nanlu06** [comment]
Sure, I tested, results are fine to me so I merged this PR. When I merge the datacard updates, it seems that I am not mering properly that some of the old commits appear in the PR. will improve next time.
